feat: update pyiceberg/catalog/hive.py to support hive 4.x.x

igorvoltaic opened the pull request:
resolves #1222

# Rationale for this change

Starting at version 4.0.1, Hive metastore removed deprecated thrift APIs that py-iceberg is currently using. When trying to create a table with catalog.create_table_transaction using Hive metastore 4.0.1, py-iceberg raise an unexpected thrift.Thrift.TApplicationException: Invalid method name: 'get_table' error

# Are there any user-facing changes?
not expected

kevinjqliu commented on pyiceberg/catalog/hive.py:
```diff
@@ -389,8 +389,8 @@ def _create_hive_table(self, open_client: Client, hive_table: HiveTable) -> None
         try:
-            return open_client.get_table(dbname=database_name, tbl_name=table_name)
-        except NoSuchObjectException as e:
+            return open_client.get_tables(db_name=database_name, pattern=table_name).pop()
```
It looks like `get_tables` is available in previous versions of thrift clients 

i see
```
    def get_tables(self, db_name, pattern):
        """
        Parameters:
         - db_name
         - pattern

        """
        self.send_get_tables(db_name, pattern)
        return self.recv_get_tables()
```
in `vendor/hive_metastore/ThriftHiveMetastore.py`

kevinjqliu reviewed (approved):
LGTM this change seems to be backwards compatible. We dont currently have a way to test out hive 4.x.x though, but we can address that separately

kevinjqliu commented:
ah looks like CI failed because we mock `.get_table` in tests 
https://grep.app/search?f.path=tests%2F&f.path.pattern=tests&f.repo.pattern=iceberg-python&q=.get_table

need to switch those to `get_tables`

kevinjqliu reviewed (commented):
great, looks like `get_table_objects_by_name` is backwards compatible too.

Could you try this PR out against a hive 4.x.x deployment? 

It seems like we are using [hive 4.0.0](https://github.com/apache/iceberg-python/blob/2d7d0892d3cd26bf27da7c53480efd542a6c75a4/dev/hive/Dockerfile#L26) in the [integration tests](https://github.com/apache/iceberg-python/blob/2d7d0892d3cd26bf27da7c53480efd542a6c75a4/dev/docker-compose-integration.yml#L89-L99). We are using the hive client [`session_catalog_hive`](https://github.com/apache/iceberg-python/blob/2d7d0892d3cd26bf27da7c53480efd542a6c75a4/tests/conftest.py#L2453-L2464) in tests
https://grep.app/search?f.repo=apache%2Ficeberg-python&q=session_catalog_hive
So im confused why the tests were not failing before

kevinjqliu commented:
oh interesting. so `get_table` was not deprecated in `4.0.0`, but rather in `4.0.1` 🤔 

see the [4.0.1 changelog](https://issues.apache.org/jira/secure/ReleaseNote.jspa?version=12354786&styleName=Text&projectId=12310843)
[HIVE-26537] - Deprecate older APIs in the HMS
https://github.com/apache/hive/pull/3599/files#diff-47ffee8549a256db9156ce4287f750674b1689362362db066010ff60319dcca0L288

kevinjqliu commented:
@igorvoltaic lets run the integration tests against 4.0.1! 
https://github.com/apache/iceberg-python/blob/ad8263b1be048c8cb67d40efe70f494a4f1cb374/dev/hive/Dockerfile#L26

kevinjqliu commented:
Looks like `get_table_objects_by_name` was removed in 4.0.1 too 
https://github.com/apache/hive/pull/3599/files#diff-47ffee8549a256db9156ce4287f750674b1689362362db066010ff60319dcca0L293
so we'll need to find a different solution.

I found this out after getting the hive 4.0.1 run with local pyiceberg, see #2217

kevinjqliu commented:
ok this the expected replacement
https://github.com/apache/hive/pull/3599/files#diff-2554b13fe8458e70912484d7e9c3f0c3403d0b27fe1affced872e576f2742ea9R268-R269

also feel free to cherry pick over #2217

Fokko commented:
Great catch @kevinjqliu:

> cant use `get_table_objects_by_name` its also removed in hive 4.0.1

To catch this, we probably want to regenerate the vendor package against 4.0.1: https://github.com/apache/iceberg-python/tree/main/vendor#vendor-packages

igorvoltaic commented:
> Great catch @kevinjqliu:
> 
> > cant use `get_table_objects_by_name` its also removed in hive 4.0.1
> 
> To catch this, we probably want to regenerate the vendor package against 4.0.1: https://github.com/apache/iceberg-python/tree/main/vendor#vendor-packages

Hey @Fokko, there is one already https://raw.githubusercontent.com/apache/hive/refs/heads/master/standalone-metastore/metastore-common/src/gen/thrift/gen-py/hive_metastore/ThriftHiveMetastore.py

> cant use `get_table_objects_by_name` its also removed in hive 4.0.1

@kevinjqliu thanks for your help with testing. will try to find a way to keep it backwards compatible, but not so sure yet

kevinjqliu commented:
ah this is the gift that keeps on giving... 

I've made a couple changes in https://github.com/apache/iceberg-python/pull/2217 to
- use hive 4.0.1 in integration tests
- use current pyiceberg instead of previous versions in integration tests
- regenerate `vendor/` 

Using hive 4.0.1 is blocked now due to https://github.com/apache/iceberg/issues/12878 because spark hms connector is not compatible

igorvoltaic commented:
I think I got this. Made some more changes and we've tested those in our stage v4 server, moving on to production. Do we want to keep backwards compatibility with v3?
If so I'll push new updates in a couple of days

kevinjqliu commented:
I was not able to get our integration tests working with hive 4.0.1, see #2217. The main issue is with spark's hive catalog using deprecate APIs, https://github.com/apache/iceberg/issues/12878

> Made some more changes and we've tested those in our stage v4 server, moving on to production. Do we want to keep backwards compatibility with v3?

As long as these changes are backwards compatible with older hive clients/servers, we can make this change. But I'd need your help to ensure these changes work against hive 4.0.1

igorvoltaic commented:
Hey 👋 @kevinjqliu! Could you give it another look? I've kept this backwards compatible

Thou it worked I've struggled for a while with tests but finally got there https://github.com/igorvoltaic/iceberg-python/actions/runs/17369350370/job/49301826288

To test 4.0.1 and onwards there might be another Dockerfile needed. In my case this is what was used just to make sure it is okay after adding new changes: 
Bumped patch in [dev/hive/Dockerfile](https://github.com/apache/iceberg-python/blob/main/dev/hive/Dockerfile#L26)
Run `docker compose -f dev/docker-compose-integration.yml up -d --build`
And run this simple script
```
import string
from random import choice

from pyiceberg.catalog.hive import HiveCatalog
from pyiceberg.schema import (
    Schema,
)
from pyiceberg.types import (
    IntegerType,
    MapType,
    NestedField,
    StringType,
    StructType,
)


def table_name() -> str:
    prefix = "my_iceberg_table-"
    random_tag = "".join(choice(string.ascii_letters) for _ in range(3))
    return (prefix + random_tag).lower()

def database_name() -> str:
    prefix = "my_iceberg_database-"
    random_tag = "".join(choice(string.ascii_letters) for _ in range(3))
    return (prefix + random_tag).lower()


test_catalog = HiveCatalog(
    "test_hive_catalog",
    **{
        "uri": "http://localhost:9083",
        "s3.endpoint": "http://localhost:9000",
        "s3.access-key-id": "admin",
        "s3.secret-access-key": "password",
    },
)


table_schema_nested = Schema(
    NestedField(
        field_id=6,
        name="id_to_person",
        field_type=MapType(
            key_id=7,
            key_type=IntegerType(),
            value_id=8,
            value_type=StructType(
                NestedField(field_id=10, name="name", field_type=StringType(), required=False),
                NestedField(field_id=11, name="age", field_type=IntegerType(), required=True),
            ),
            value_required=True,
        ),
        required=True,
    ),
    schema_id=1,
    identifier_field_ids=[],
)

identifier = (database_name(), table_name())
test_catalog.create_namespace(identifier[0])
test_catalog.create_table(identifier, table_schema_nested)
table = test_catalog.load_table(identifier)
test_catalog.drop_table(identifier)
test_catalog.drop_namespace(identifier[0])

assert table.name() == identifier
```
